### ISSUE TITLE
Map/List support for Consul_Keys

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-consul#15]
_This issue was originally opened by @wv-cfromm as hashicorp/terraform#15356. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

My project is a self-serve build/compile/deploy on immutable servers.  In numerous places, I have a need for complex structures as variables, such as a structure to hold ip/port manifest for binding and security group creation.  Maps/Lists (or nesting both) works fine as cli vars or tfvars, but Consul KV retrieved with data.consul_keys is all String.  This is rough because I need Consul to carry my vars from other apps into TF.

Am I missing something? Is there some way to interpret KV as other than String?

If this needs to be a feature request, I believe there is a number of approaches.  Interpreting Consul KV as other than string is one, such as storing HCL in Consul .  It might be better to enhance the various interpolation functions to support accepting a single string input.  For example, map function only works when supplied key/value, key/value, etc.   However, the map function could accept a single string that includes a map structure. Same goes for List, etc. Or separate function that cast a string into map, list, etc assuming standard map/list/etc syntax.  Or what about accepting HCL/JSON as input?

**Comment by mattlqx** [comment]
I have a similar use-case and I would also desire native support for lists, if not both. I've tried using a `null_data_source` to split the a comma-delimited list, but it has the same string-only limitations. I think the way that you can currently obtain what you want is writing a module that will take a Consul path and output the result of a split based on whatever delimiter. But like you, I'd rather these data sources simply support something more complex than just strings.

**Comment by wv-cfromm** [comment]
Yes, thats what I am doing today, but it quickly becomes unbearable.  

For example "elb:0.0.0.0/0=80,443;app:sg-elb=80,81" with the order of splits being on ;:|=.  And this is a simple one.

data.consul_keys.var should have the same capabilities as -var on command line.

**Comment by apparentlymart** [comment]
Hi @wv-cfromm! Sorry for the slow response here.

Could you share an example of the set of keys you're trying to write into Consul? I'm not quite sure what is the best behavior here since Consul _itself_ only supports storing strings, and so Terraform is attempting to follow the same abstraction level as the underlying Consul API.

We are currently in the early stages of planning some configuration language improvements that I think might help here:

* A more general `jsonencode` function that can accept arbitrary data structures to encode, rather than right now only supporting maps with string values. This would allow packing a complex data structure into a _single_ consul key.
* A way to perform arbitrary transformations on the items in a list or map to produce a new list or map. That could, for example, help to express the flattening of a data structure into the flat string map the Consul's key/value store expects, allowing a data structure to be spread over _multiple_ Consul keys in a user-defined way.

...but I'm not totally sure I completely understand what you want to achieve here, so having a real example to look at would help to see if anything else is needed here.

**Comment by mattlqx** [comment]
+1 on the `jsonencode` function. It's always seemed odd to me that you can only take in JSON and not output it. That would at least give symmetry to that situation.

**Comment by wv-cfromm** [comment]
In my principle use case, I have a genericised TF config for an autoscaling group (and associated resources, security group, loadbalancer, etc), but its written to be dynamic and support different builds based on variables.    One of those variables is the structure for IPs/Ports/Protocols, used to build LB bindings and security group rules.  

Because consul_keys are only interpretted as string, my only option is a multi-delimited string.  For example, I am currently retreiving "elb:0.0.0.0/0=80,443;app:sg-elb=80,81" from consul_keys and then splitting with the order of splits being on ;:|=.  This essentially splits out into two rule lists, one for ELB and one for APP.  That subsquently splits into IP/SG allowance and ports.  This is pretty ugly.

This is a bit easier in CLI, as I can simplify the structure significantly with map of lists and/or list of maps.  While the value in put in CLI, is technically a string, i.e. 'items={item1 = "1",item2 = "2",item3 = "3"}', TF interprets this as a map due to agreed syntax and variable definition.

My basic ask is that values output from consul_keys gets the same treatment.  If the value from consul_keys is 'items={item1 = "1",item2 = "2",item3 = "3"}', then TF should recognize this as a map and treat it accordingly. If the value from consul_keys is 'items=["1","2","3"]', then TF should recognize this as a list and treat it accordingly. Same for nesting.

I do agree that the capability of a ~payload~ variable, like JSON, would be a huge improvement for TF as a whole.  However, it seems you are talking about input into Consul.  You reference jsonencode and packing structure into a key.  While it would be awesome if it goes both ways, I'm mostly concerned with output.  I want to store a JSON structure in consul (or pass from CLI/static), and have TF retrieve that structure AND allow referencing its nodes.

```
# CONSUL KEY SGMap
{
  "elb": {
    "security": [
      {
        "src": "34.53.12.4/32",
        "dst": "80"
      },
      {
        "src": "34.53.12.4/32",
        "dst": "443"
      },
      {
        "src": "28.24.21.2/32",
        "dst": "80"
      },
      {
        "src": "28.24.21.2/32",
        "dst": "443"
      },
      {
        "src": "36.53.23.1/32",
        "dst": "80"
      },
      {
        "src": "36.53.23.1/32",
        "dst": "443"
      }
    ],
    "bindings": [
      {
        "src": "80",
        "dst": "80"
      },
      {
        "src": "443",
        "dst": "81"
      }
    ]
  }
  "app": {
    "security": [
      {
        "src": "sg-e55555e",
        "dst": "80"
      },
      {
        "src": "sg-e55555e",
        "dst": "81"
      },
      {
        "src": "192.168.0.0/24",
        "dst": "80"
      },
      {
        "src": "192.168.0.0/24",
        "dst": "443"
      }
    ]
  }
}
```
I should then be able to use node references in TF.   

```
data "consul_keys" "sgmap" {
    key {
        name = "sgmap"
        path = "/sgmap"
    }
resource "aws_security_group_rule" "lb_ingress1" {
    type = "ingress"
    from_port = "${data.consul_keys.sgmap.var.sgmap.elb.security[countindex].dst}"
    to_port = "${data.consul_keys.sgmap.var.sgmap.elb.security[countindex].dst}"
    protocol = "tcp"
    cidr_blocks = ["${data.consul_keys.sgmap.var.sgmap.elb.security[countindex].src}"]
    security_group_id = "${aws_security_group.sg_lb.id}"
}
```
Or static CLI
```
variable "sgmap" {
    type = "json"
}
resource "aws_security_group_rule" "lb_ingress1" {
    type = "ingress"
    from_port = "${var.sgmap.elb.security[countindex].dst}"
    to_port = "${var.sgmap.elb.security[countindex].dst}"
    protocol = "tcp"
    cidr_blocks = ["${var.sgmap.elb.security[countindex].src}"]
    security_group_id = "${aws_security_group.sg_lb.id}"
}
```
The references from the JSON elements could also be referenced using some jsondecode function.   `"${jsondecode(data.consul_keys.sgmap.var.sgmap, elb.security[countindex].dst}"` and `"${jsondecode(var.sgmap, elb.security[countindex].dst}"`

**Comment by apparentlymart** [comment]
Oh, indeed... a `jsondecode` function is something we're looking at too. I was just focused on the encoding case because your issue seemed to be primarily about _writing_ to Consul. However, looking back now I see that you were probably talking about _reading_ from Consul instead.

Generally-speaking Terraform's design errs on the side of being explicit rather than doing "magical" inferences, since this reduces surprises. However, if we can define this in such a way that it's intuitive and explicit then I think that could be fine... my gut says that explicitly using more-advanced `jsonencode` and `jsondecode` functions is the answer here, along with some other configuration language changes that might result in it looking something like this:

```hcl
# This is a DRAFT, and not currently implemented. Final design may differ from this example.

# Example of *writing* values to Consul
resource "consul_keys" "write" {
  key {
    path = "example"
    value = jsonencode([
      {
        src = "192.168.0.0/24"
        dst = "80"
      },
      {
        src = "192.168.0.0/24"
        dst = "443"
      },
    ])
  }
}

# Example of *reading* from Consul
# (presumably in practice this would be a separate config)
data "consul_keys" "read" {
  key {
    name = "example"
    path = "example"
  }
}

# Using the "local value" feature from PR #15449, which is not yet merged
locals {
  security_group_rules = jsondecode(data.consul_keys.read["example"])
}

resource "aws_security_group_rule" "lb_ingress1" {
  count             = length(local.security_group_rules)
  type              = "ingress"
  from_port         = local.security_group_rules[count.index].dst
  to_port           = local.security_group_rules[count.index].dst
  protocol          = "tcp"
  cidr_blocks       = [local.security_group_rules[count.index].src]
  security_group_id = aws_security_group.sg_lb.id
}
```

**Comment by wv-cfromm** [comment]
Ummm Yes, please please please.  Where to send bribes?

**Comment by spanktar** [comment]
I came across this trying to WRITE to Consul and save myself a lot of repetition.  For example:

```
variable "druid_historical" {
  default = {
    cache.bytes        = "128000000"
    cache.column.bytes = "64000000"
    cache.segment.max  = "80000000000"
    heap               = "4g"
    memory.direct      = "2400m"
    memory.new         = "512m"
    port               = "8083"
    proc.buffer        = "128000000"
    proc.threads       = ""
    threads            = "3"
    tier.priority      = ""
    tier.storage       = ""
  }
}

resource "consul_keys" "consul_keys_druid" {
  key {
    path  = "druid/historical"
    value = "${var.druid_historical}"
  }
}
```

What I would hope for here is that the map keys are added under the path as a root, resulting in the following Consul keys:

```
druid/historical/cache.bytes
druid/historical/cache.column.bytes
druid/historical/cache.segment.max
etc...
```

As it stands now, I have to write each one out. It's do-able, just boring ;)

**Comment by apparentlymart** [comment]
Hi @spanktar,

I think what you're describing there is what [the `consul_key_prefix` resource](https://www.terraform.io/docs/providers/consul/r/key_prefix.html) does. Passing in your `var.druid_historical` to its `subkeys` argument should get the result you're looking for.

**Comment by apparentlymart** [comment]
[The `jsondecode` function](https://www.terraform.io/docs/configuration/functions/jsondecode.html) I mentioned before is now available for testing in [v0.12.0-beta1](https://www.hashicorp.com/blog/announcing-terraform-0-1-2-beta1), and will be included in the forthcoming v0.12.0 final release. This also includes a generalization of [the `jsonencode` function](https://www.terraform.io/docs/configuration/functions/jsonencode.html) that no longer requires all of the values to be strings.

I think this, combined with the `consul_key_prefix` resource type and data source, cover the use-cases discussed here. More complicated transforms can be achieved using [`for` expressions](https://www.terraform.io/docs/configuration/expressions.html#for-expressions) (also added in v0.12.0) to generate the `consul_key_prefix` map by transforming another data structure.

Since Terraform Core has these features all merged in the master branch and ready for release, I'm going to close this out. The Consul provider doesn't yet have a stable release compatible with v0.12 but that will follow shortly. In the mean time, you can use the dev snapshots described in the beta announcement blog post for testing. Thanks everyone for sharing your use-cases here.
